### PR TITLE
feat: Add traffic streams

### PR DESCRIPTION
**Pull request by sicarul**
This PR adds the Traffic streams to collect how much traffic did we have on each Repository.

It'll only collect data for the repositories the token has write access to, and will log an info message for those that it couldn't collect.

### CODE REVIEW

File: tap_github/repository_streams.py [hunk 1]
```diff
@@ -2237,15 +2238,12 @@ def validate_response(self, response: requests.Response) -> None:
         except FatalAPIError as e:
             if "Must have push access to repository" in str(e):
                 self.logger.info(
-                    (
-                        f"We do not have permissions for this repository: e \t"
-                    )
+                    (f"We do not have permissions for this repository: e \t")
```
[the remark this answers is not in the document]
Comment (sicarul) [reply tap_github/repository_streams.py]: Applied!

**Comment by sicarul** [comment]
It seems the test failed:

```

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tap_github/tests/test_tap.py::test_get_a_repository_in_repo_list_mode[False] - singer_sdk.exceptions.FatalAPIError: 403 Client Error: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/metrics/traffic#get-repository-clones"}' (Reason: Forbidden) for path: /repos/MeltanoLabs/tap-github/traffic/clones
FAILED tap_github/tests/test_tap.py::test_get_a_repository_in_repo_list_mode[True] - singer_sdk.exceptions.FatalAPIError: 403 Client Error: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/metrics/traffic#get-repository-clones"}' (Reason: Forbidden) for path: /repos/MeltanoLabs/tap-github/traffic/clones
FAILED tap_github/tests/test_tap.py::test_last_state_message_is_valid - singer_sdk.exceptions.FatalAPIError: 403 Client Error: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/metrics/traffic#get-repository-clones"}' (Reason: Forbidden) for path: /repos/MeltanoLabs/tap-github/traffic/clones
============= 3 failed, 7 passed, 714 warnings in 84.81s (0:01:24) =============
Error: Process completed with exit code 1.
```

I think it's because of a permissions issue on the user running the test.

**Comment by edgarrmondragon** [comment]
> It seems the test failed:
> 
> ```
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> =========================== short test summary info ============================
> FAILED tap_github/tests/test_tap.py::test_get_a_repository_in_repo_list_mode[False] - singer_sdk.exceptions.FatalAPIError: 403 Client Error: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/metrics/traffic#get-repository-clones"}' (Reason: Forbidden) for path: /repos/MeltanoLabs/tap-github/traffic/clones
> FAILED tap_github/tests/test_tap.py::test_get_a_repository_in_repo_list_mode[True] - singer_sdk.exceptions.FatalAPIError: 403 Client Error: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/metrics/traffic#get-repository-clones"}' (Reason: Forbidden) for path: /repos/MeltanoLabs/tap-github/traffic/clones
> FAILED tap_github/tests/test_tap.py::test_last_state_message_is_valid - singer_sdk.exceptions.FatalAPIError: 403 Client Error: b'{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/metrics/traffic#get-repository-clones"}' (Reason: Forbidden) for path: /repos/MeltanoLabs/tap-github/traffic/clones
> ============= 3 failed, 7 passed, 714 warnings in 84.81s (0:01:24) =============
> Error: Process completed with exit code 1.
> ```
> 
> I think it's because of a permissions issue on the user running the test.

@sicarul Yeah, can you add note to the readme documenting which permissions are required to pull the traffic streams?

**Comment by sicarul** [comment]
Added @edgarrmondragon !

**Comment by sicarul** [comment]
@edgarrmondragon do we need anything else to move this forward?

**Comment by edgarrmondragon** [comment]
> @edgarrmondragon do we need anything else to move this forward?

@sicarul Tests are still failing so I worry that this may cause similar failures to users of the tap if their tokens don't support pulling from the traffic endpoints.

Wdyt about adding a tap setting (e.g. `sync_traffic_streams`) that by default makes the tap omit these streams?

cc @kgpayne if you have other ideas

**Comment by sicarul** [comment]
> > @edgarrmondragon do we need anything else to move this forward?
> 
> @sicarul Tests are still failing so I worry that this may cause similar failures to users of the tap if their tokens don't support pulling from the traffic endpoints.
> 
> Wdyt about adding a tap setting (e.g. `sync_traffic_streams`) that by default makes the tap omit these streams?
> 
> cc @kgpayne if you have other ideas

Maybe i could also handle this error instead?
“ Resource not accessible by integration”

However, i do not believe any user would try to run this entire tap without selecting streams, some are monstruous.

**Comment by edgarrmondragon** [comment]
> However, i do not believe any user would try to run this entire tap without selecting streams, some are monstruous.

@sicarul You're right, but the SDK tries to be helpful to folks just getting started with singer.io and so all streams are selected by default if no catalog is passed.

However, we could change that in your streams. I'll propose some changes.

(I also started a discussion in the SDK: https://github.com/meltano/sdk/discussions/1652)

File: tap_github/repository_streams.py [hunk 2]
```diff
@@ -2219,3 +2219,126 @@ def query(self) -> str:
...
+        """
+        try:
+            super().validate_response(response)
+        except FatalAPIError as e:
+            if "Must have push access to repository" in str(e):
+                self.logger.info(
+                    "We do not have permissions for this repository", exc_info=True
+                )
+                return
+            raise
```
Comment (edgarrmondragon) [inline tap_github/repository_streams.py:2247]:
```suggestion
        if response.status_code == 403:
            contents = response.json()
            if contents["message"] == "Resource not accessible by integration":
                self.logger.info("Permissions missing to sync stream '%s'", self.name)
                return
        super().validate_response(response)
```

File: tap_github/repository_streams.py [hunk 3]
```diff
@@ -2219,3 +2219,126 @@ def query(self) -> str:
             ),
         ),
     ).to_dict()
+
+
+class TrafficRestStream(GitHubRestStream):
+    """Base class for Traffic Streams"""
+
```
Comment (edgarrmondragon) [inline tap_github/repository_streams.py:2226]:
```suggestion
    """Base class for Traffic Streams"""

    @property
    def metadata(self):
        """Override default selection metadata for this stream.
        
        TODO: Remove this in favor of the recommended approach when the SDK has one.
        """
        result = super().metadata
        if self._tap_input_catalog is None:
            result.root.selected = False
        return result

```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]

File: tap_github/repository_streams.py [hunk 4]
```diff
@@ -2219,3 +2219,126 @@ def query(self) -> str:
...
+            return []
+
+        """Parse the response and return an iterator of result rows."""
+        yield from extract_jsonpath(self.records_jsonpath, input=response.json())
+
+    def validate_response(self, response: requests.Response) -> None:
+        """Allow some specific errors.
+        Do not raise exceptions if the error says "Must have push access to repository"
+        as we actually expect these in this stream when we don't have write permissions into it.
+        """
```
[the remark this answers is not in the document]
Comment (sicarul) [reply tap_github/repository_streams.py:2276]: Accepted!

File: tap_github/repository_streams.py [hunk 5]
```diff
@@ -2246,3 +2246,134 @@ def query(self) -> str:
...
+
+class TrafficRestStream(GitHubRestStream):
+    """Base class for Traffic Streams"""
+
+    @property
+    def metadata(self):
+        """Override default selection metadata for this stream.
+        
+        TODO: Remove this in favor of the recommended approach when the SDK has one.
+        """
```
Comment (edgarrmondragon) [inline tap_github/repository_streams.py:2259]:
```suggestion
        """Override default selection metadata for this stream.

        TODO: Remove this in favor of the recommended approach when the SDK has one.
        """
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -71,6 +71,10 @@ To avoid this, the GitHub streams will exit early. I.e. when there are no more `
 
 You can easily run `tap-github` by itself or in a pipeline using [Meltano](www.meltano.com).
 
+### Notes regarding permissions
+
+* For the traffic_* streams, [you will need write access to the repository](https://docs.github.com/en/rest/metrics/traffic?apiVersion=2022-11-28)
```
Comment (edgarrmondragon) [inline README.md:76]:
```suggestion
* For the `traffic_*` streams, [you will need write access to the repository](https://docs.github.com/en/rest/metrics/traffic?apiVersion=2022-11-28). You can enable extraction for these streams by [selecting them in the catalog](https://hub.meltano.com/singer/spec/#metadata).
```

**Comment by edgarrmondragon** [comment]
@sicarul this LGTM! Just a minor comment on the readme and this is good to merge for me 🙂

**Review by laurentS — APPROVED** [review]

**Comment by edgarrmondragon** [comment]
Thanks @sicarul!

**Pull request merged**